router-tests/testenv/testenv.go

```diff
@@ -1910,68 +1914,176 @@ func (e *Environment) WaitForTriggerCount(desiredCount uint64, timeout time.Dura
 	}
 }
 
-func DeflakeWSReadMessage(t testing.TB, conn *websocket.Conn) (messageType int, p []byte, err error) {
-	for i := 0; i < 5; i++ {
+func WSReadMessage(t testing.TB, conn *websocket.Conn) (messageType int, p []byte, err error) {
+	b := backoff.New(5*time.Second, 100*time.Millisecond)
+
+	attempts := 0
+	maxAttempts := 5
+
+	for attempts < maxAttempts {
+		attempts++
+
+		// Set a reasonable read deadline
+		readDeadline := time.Now().Add(2 * time.Second)
+		if err = conn.SetReadDeadline(readDeadline); err != nil {
+			t.Logf("Failed to set read deadline: %v", err)
+			return 0, nil, err
+		}
+
 		messageType, p, err = conn.ReadMessage()
-		if err != nil && strings.Contains(err.Error(), "connection reset by peer") {
-			t.Log("connection reset by peer found, retrying...")
-			err = conn.SetReadDeadline(time.Now().Add(1 * time.Second))
-			require.NoError(t, err)
-			time.Sleep(time.Duration(i*200) * time.Millisecond)
-			continue
+
+		// Reset the deadline to prevent future operations from timing out
+		if resetErr := conn.SetReadDeadline(time.Time{}); resetErr != nil {
+			t.Logf("Failed to reset read deadline: %v", resetErr)
+		}
+
+		if err == nil {
+			return messageType, p, nil
 		}
-		break
+
+		if attempts >= maxAttempts {
+			return 0, nil, fmt.Errorf("failed to read from WebSocket after %d attempts: %w", attempts, err)
+		}
+
+		// Get backoff duration for this attempt
+		duration := b.Duration()
+		t.Logf("WebSocket read error (attempt %d/%d): %v - retrying in %v...",
+			attempts, maxAttempts, err, duration)
+
+		time.Sleep(duration)
 	}
 
-	return messageType, p, err
+	// This should not be reached
+	return 0, nil, fmt.Errorf("failed to read from WebSocket: %w", err)
 }
 
-func DeflakeWSReadJSON(t testing.TB, conn *websocket.Conn, v interface{}) (err error) {
-	for i := 0; i < 5; i++ {
+func WSReadJSON(t testing.TB, conn *websocket.Conn, v interface{}) (err error) {
```

Comment:
If these don't necessarily need exponential backoff, perhaps [`require.EventuallyWithT`](https://pkg.go.dev/github.com/stretchr/testify/require#EventuallyWithT) would do and make the helper function less complex?

Reply:
Good thoughts. The idea here was to make it more reliable. I think your idea could also be enough.

Reply:
Eventually works with a callback that makes writing it harder.